crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -0,0 +1,505 @@
+            format_comment(format_off_comment)
+        ]
+    )?;
+
+    format_off_comment.mark_formatted();
+
+    // Now inside a suppressed range
+    write_suppressed_statements(
+        format_off_comment,
+        first_suppressed,
+        leading_comment_ranges.as_slice(),
+        statements,
+        f,
+    )
+}
+
+/// Disables formatting for all statements between the `last_formatted` and the first trailing or leading `fmt: on` comment.
+/// The statements are formatted as they appear in the source code.
+///
+/// Returns the last formatted statement.
+///
+/// ## Panics
+/// If `last_formatted` has no trailing suppression comment.
+#[cold]
+pub(crate) fn write_suppressed_statements_starting_with_trailing_comment<'a>(
+    last_formatted: SuiteChildStatement<'a>,
+    statements: &mut std::slice::Iter<'a, Stmt>,
+    f: &mut PyFormatter,
+) -> FormatResult<&'a Stmt> {
+    let comments = f.context().comments().clone();
+    let source = f.context().source();
+
+    let trailing_node_comments = comments.trailing_comments(last_formatted);
+    let mut trailing_comment_ranges =
+        CommentRangeIter::outside_suppression(trailing_node_comments, source);
+
+    // Formatted comments gets formatted as part of the statement.
+    let (_, mut format_off_comment) = trailing_comment_ranges
+        .next()
+        .expect("Suppressed statement to have trailing comments")
+        .unwrap_suppression_starts();
+
+    let maybe_suppressed = trailing_comment_ranges.as_slice();
+
+    // Mark them as formatted so that calling the node's formatting doesn't format the comments.
+    for comment in maybe_suppressed {
+        comment.mark_formatted();
+    }
+    format_off_comment.mark_formatted();
+
+    // Format the leading comments, the node, and the trailing comments up to the `fmt: off` comment.
+    last_formatted.fmt(f)?;
+
+    format_off_comment.mark_unformatted();
+    TrailingFormatOffComment(format_off_comment).fmt(f)?;
+
+    for range in trailing_comment_ranges {
+        match range {
+            // A `fmt: off`..`fmt: on` sequence. Disable formatting for the in-between comments and
+            SuppressionComments::SuppressionEnds {
+                suppressed_comments: _,
+                format_on_comment,
+                formatted_comments,
+                format_off_comment: new_format_off_comment,
+            } => {
+                format_on_comment.mark_unformatted();
+
+                for comment in formatted_comments {
+                    comment.mark_unformatted();
+                }
+
+                write!(
+                    f,
+                    [
+                        verbatim_text(TextRange::new(
+                            format_off_comment.end(),
+                            format_on_comment.start(),
+                        )),
+                        trailing_comments(std::slice::from_ref(format_on_comment)),
+                        trailing_comments(formatted_comments),
```

Comment:
Could it ever be the case that we're missing out on "special" formatting of some of these comments? I don't have a more precise idea of when that could happen, but just acknowledging that these are typically handled by children, and we're now intercepting and formatting them here.

Reply:
Good catch. This could be a problem if a statement overrides `fmt_leading_comments` or `fmt_trailing_comments`. I checked, and no expression or statement overrides these methods on `FormatNodeRule`. I went ahead and inlined the functions to not give the impression that this behavior should be customized. 

Either way, I think it would be okay to format the comments with the default formatting. It may not be perfect but we're dealing with suppressed ranges anyway.